### DISCUSSION TITLE
ITypeInfo & void interface members

### DISCUSSION THREAD

**Discussion by Greedquest** [twinbasic/twinbasic#1257] [Q&A]
I'm trying to write up ITypeInfo in tB.

```vba
Interface ITypeInfo Extends IUnknown
    
    'HRESULT  GetTypeAttr([out] TYPEATTR **ppTypeAttr )
    Sub GetTypeAttr(ByRef outpTypeAttr As LongPtr)
    Sub GetTypeComp()
    Sub GetFuncDesc()
    Sub GetVarDesc()
    Sub GetNames()
    Sub GetRefTypeOfImplType()
    Sub GetImplTypeFlags()
    Sub GetIDsOfNames()
    Sub Invoke()
    'HRESULT  GetDocumentation( [in] MEMBERID memid, [out] BSTR *pBstrName, [out] BSTR *pBstrDocString, [out] DWORD *pdwHelpContext, [out] BSTR *pBstrHelpFile)
    Sub GetDocumentation(ByVal memid As DISPID, ByRef outName As String, Optional ByVal pBstrDocString As LongPtr = NULL_PTR, Optional ByVal pdwHelpContext As LongPtr = NULL_PTR, Optional ByVal pBstrHelpFile As LongPtr = NULL_PTR)
    Sub GetDllEntry()
    Sub GetRefTypeInfo()
    Sub AddressOfMember()
    Sub CreateInstance()
    Sub GetMops()
    Sub GetContainingTypeLib()
    'void ITypeInfo::ReleaseTypeAttr( [in] TYPEATTR *pTypeAttr)
    Sub ReleaseTypeAttr(ByVal pTypeAttr As LongPtr)
    Sub ReleaseFuncDesc()
    Sub ReleaseVarDesc()
End Interface
```

I have some questions, I think this is a good place to ask but I can ask on Stack Overflow too. 

1) I only need a few members so I'm wondering, is it possible only to declare the members I want to use:

```vba
Interface ITypeInfo Extends IUnknown
    [VTableIndex(0)]
    Sub GetTypeAttr(ByRef outpTypeAttr As LongPtr)
    [VTableIndex(9)]
    Sub GetDocumentation(ByVal memid As DISPID, ByRef outName As String, Optional ByVal pBstrDocString As LongPtr = NULL_PTR, Optional ByVal pdwHelpContext As LongPtr = NULL_PTR, Optional ByVal pBstrHelpFile As LongPtr = NULL_PTR)
    [VTableIndex(16)]
    Sub ReleaseTypeAttr(ByVal pTypeAttr As LongPtr)
End Interface
```

Or mark the other members as stubs/not-invokable to the compiler.

2) `HRESULT` subs are the default so I assume they are declared as I have. But what about a void method (`Sub ReleaseTypeAttr(ByVal pTypeAttr As LongPtr)` is this going to lead to a misaligned stack since there is no HRESULT to pop

3) We don't have a way to express an out-TYPEATTR (in GetTypeAttr) struct, I have to use copy memory and pointers IIUC

**Comment by fafalone** [comment]
1) Why not just add Sub Dummy1() Sub Dummy2() etc? That's what would have to be done behind the scenes I believe anyway.

2) The new [ PreserveSig ] attribute can be used for that.
```
[ PreserveSig ]
Sub ReleaseTypeAttr(ByVal pTypeAttr As LongPtr)
```

3) It's unusual in that it's **UDT... obviously there's normally no issue with [out] *UDT

**Reply by Greedquest** [reply]
@fafalone Thanks Dummy123 is a good way to go about it, I was worried about accidentally calling the methods which are just stubs and have no proper signatures which would crash stuff. There is a `[Restricted]` attribute might also help the compiler enforce these dummy methods aren't called. Still it's a lot of visual noise.

 Is that what `[PreserveSig]` is for?! Never realised. It should be called something like `[NoHRESULT]` or just `[void]/[VoidMethod]` what exactly is it "preserving"?

Yeah the UDT one is weird, it's because the callee owns the memory so you the caller have to tell them to deallocate it when you are done.

**Reply by bclothier** [reply]
> Is that what `[PreserveSig]` is for?! Never realised. It should be called something like `[NoHRESULT]` or just `[void]/[VoidMethod]` what exactly is it "preserving"?

Note that it's based on the [C#'s version](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.preservesigattribute?view=net-7.0). 

As the example shows, calling it `NoHResult` or `void` would not really make sense because you can define a method that returns a long that's not a HRESULT but there's no way to know the difference.

**Reply by WaynePhillipsEA** [reply]
> Is that what `[PreserveSig]` is for?! Never realised. It should be called something like `[NoHRESULT]` or just `[void]/[VoidMethod]` what exactly is it "preserving"?

It's preserving the signature of the method, as you've defined it, rather than embellishing it with HRESULT and moving the return value into the last ([out, retval]) parameter which is what normally happens.   As it's not a feature that's needed all that often, I feel using the same named attribute as .NET is appropriate.

**Reply by wqweto** [reply]
TB can now be marketed as VB6 with attributes :-))

**Reply by fafalone** [reply]
Most commonly I use it to handle the HRESULT manually. Like IEnumIDList::Next *does* return an HRESULT, but a nonzero HRESULT is a normal, every-time occurrence when it's out of items to enumerate, and I need to know what that happens to exit the loop... so I don't want VB's special handling to raise an error jumping out to a handler or ignoring it. There's a lot of others like this, where you want the HRESULT value, but don't want the VB handler.
